File: day4/day4.py

import itertools
import re


def year_range(low, high):
    """Returns a function that tests whether year is in range, inclusive"""
    return lambda year_str: low <= int(year_str) <= high

def matches_pattern(pattern):
    return lambda x: pattern.match(x) is not None

PID_PATTERN = re.compile(r"^\d{9}$")
HCL_PATTERN = re.compile(r"^#[0-9a-f]{6}$")

def valid_height(x):
    match = re.match(r"^(\d+)(in|cm)$", x)
    if not match:
        return False
    value = int(match.group(1))
    if match.group(2) == "in":
        return 59 <= value <= 76
    else:
        return 150 <= value <= 193

VALIDATORS = {
    "byr": year_range(1920, 2002),
    "iyr": year_range(2010, 2020),
    "eyr": year_range(2020, 2030),
    "hgt": valid_height,
    "hcl": matches_pattern(HCL_PATTERN),
    "ecl": lambda x: x in {"amb", "blu", "brn", "gry", "grn", "hzl", "oth"},
    "pid": matches_pattern(PID_PATTERN),
}

def is_field_valid(field_name, value):
    """
    Returns True iff the value is valid for the given field
    """
    return VALIDATORS[field_name](value)


class Passport:
    """
    Holds the key/value mapping for one passport.
    """

    def __init__(self, text):
        """
        Parses one passport.  Input should be a block of text
        with key:value pairs separated by spaces or newlines.
        """
        self.data = dict(
            item.split(":")
            for item in text.split()
        )

    def has_all_fields(self, field_names):
        return field_names <= self.field_names

    def all_fields_valid(self, field_names):
        return all(
            field_name in self.field_names and is_field_valid(field_name, self.data[field_name])
            for field_name in field_names
        )

    @property
    def field_names(self):
        return set(self.data.keys())


def is_blank(line:str):
    return line.strip() == ""


def parse_file(text:str):
    """
    Parses a file containing multiple passports, returning a list
    of Passport objects.
    """
    return [
        Passport("".join(group))
        for key, group in itertools.groupby(text.splitlines(keepends=True), is_blank)
        if not key
    ]

# Count the valid passports in the input file, not requiring the "cid" field.
required_fields = {
    "byr",
    "iyr",
    "eyr",
    "hgt",
    "hcl",
    "ecl",
    "pid",
    # "cid",
}
with open("input.txt") as f:
    input_text = f.read()
print(
    "Day 4, part 1:",
    sum(
        1
        for passport in parse_file(input_text)
        if passport.has_all_fields(required_fields)
    )
)
print(
    "Day 4, part 2:",
    sum(
        1
        for passport in parse_file(input_text)
        if passport.all_fields_valid(required_fields)
    )
)

File: elixir_advent/lib/day4.ex
defmodule Day4 do
  def run(lines) do
    with input = parse_input(lines),
         part1 = Enum.count(input, &part1_valid/1),
         part2 = Enum.count(input, &part2_valid/1) do
      "part1: #{part1}   part2: #{part2}"
    end
  end

  def part1_valid(passport) do
    case passport do
      %{byr: _, iyr: _, eyr: _, hgt: _, hcl: _, ecl: _, pid: _} ->
        true

      _ ->
        false
    end
  end

  def part2_valid(passport) do
    case passport do
      %{byr: byr, iyr: iyr, eyr: eyr, hgt: hgt, hcl: hcl, ecl: ecl, pid: pid} ->
        number_in_range(byr, ~r/^([0-9]{4})$/, 1920, 2002) and
          number_in_range(iyr, ~r/^([0-9]{4})$/, 2010, 2020) and
          number_in_range(eyr, ~r/^([0-9]{4})$/, 2020, 2030) and
          height_valid(hgt) and
          String.match?(hcl, ~r/^#[0-9a-f]{6}$/) and
          Enum.member?(~w{amb blu brn gry grn hzl oth}, ecl) and
          String.match?(pid, ~r/^[0-9]{9}$/)

      _ ->
        false
    end
  end

  def height_valid(hgt) do
    number_in_range(hgt, ~r/^([0-9]+)cm$/, 150, 193) or
      number_in_range(hgt, ~r/^([0-9]+)in$/, 59, 76)
  end

  @doc """
  Checks that the string matches the pattern, and the first capture
  group in a number in the given range (inclusive).

  ## Examples:

    iex> Day4.number_in_range("a47b", ~r/^a([47]+)b$/, 0, 50)
    true
    iex> Day4.number_in_range("a77b", ~r/^a([47]+)b$/, 0, 50)
    false
    iex> Day4.number_in_range("49", ~r/([0-9]+)/, 50, 50)
    false
    iex> Day4.number_in_range("50", ~r/([0-9]+)/, 50, 50)
    true
    iex> Day4.number_in_range("51", ~r/([0-9]+)/, 50, 50)
    false

  """
  def number_in_range(text, pattern, low, high) do
    case Regex.run(pattern, text) do
      nil ->
        false

      [_, num_str] ->
        with num = String.to_integer(num_str) do
          low <= num and num <= high
        end
    end
  end

  @doc """
  Takes the textual input an converts it into a list
  of maps.  Each map represents one passport, mapping
  field name to value.
  """
  def parse_input(lines) do
    lines
    |> AdventUtil.split_at_blank_lines()
    |> Enum.map(&parse_one_passport/1)
  end

  def parse_one_passport(lines) do
    for line <- lines,
        item <- String.split(line),
        [field, value] = String.split(item, ":"),
        into: %{} do
      {String.to_atom(field), value}
    end
  end
end
